File: src/main/java/de/spries/fleetcommander/model/facade/PlayerSpecificUniverse.java
package de.spries.fleetcommander.model.facade;

import java.util.Collection;
import java.util.List;

import de.spries.fleetcommander.model.core.Player;
import de.spries.fleetcommander.model.core.universe.Planet;
import de.spries.fleetcommander.model.core.universe.ShipFormation;
import de.spries.fleetcommander.model.core.universe.Universe;

import static de.spries.fleetcommander.model.facade.TestMode.TEST_MODE;

public class PlayerSpecificUniverse {

	private Universe originalUniverse;
	private Player viewingPlayer;

	public PlayerSpecificUniverse(Universe originalUniverse, Player viewingPlayer) {
		this.originalUniverse = originalUniverse;
		this.viewingPlayer = viewingPlayer;
	}

	public List<PlayerSpecificPlanet> getPlanets() {
		return PlayerSpecificPlanet.convert(originalUniverse.getPlanets(), viewingPlayer);
	}

	public PlayerSpecificPlanet getPlanet(int planetId) {
		return PlayerSpecificPlanet.convert(originalUniverse.getPlanetForId(planetId), viewingPlayer);
	}

	public PlayerSpecificPlanet getHomePlanet() {
		Planet homePlanet = originalUniverse.getHomePlanetOf(viewingPlayer);
		if (homePlanet != null) {
			return PlayerSpecificPlanet.convert(homePlanet, viewingPlayer);
		}
		return null;
	}

	public void sendShips(int shipCount, int originPlanetId, int destinationPlanetId) {
		originalUniverse.sendShips(shipCount, originPlanetId, destinationPlanetId, viewingPlayer);
	}

	public Collection<ShipFormation> getTravellingShipFormations() {
		if(TEST_MODE)
			return originalUniverse.getTravellingShipFormations();
		return ShipFormation.filterByCommander(originalUniverse.getTravellingShipFormations(), viewingPlayer);
	}

	protected static PlayerSpecificUniverse convert(Universe universe, Player viewingPlayer) {
		return new PlayerSpecificUniverse(universe, viewingPlayer);
	}

}

File: src/main/kotlin/de/spries/fleetcommander/model/facade/PlayerSpecificUniverse.kt
package de.spries.fleetcommander.model.facade

import de.spries.fleetcommander.model.core.Player
import de.spries.fleetcommander.model.core.universe.ShipFormation
import de.spries.fleetcommander.model.core.universe.Universe


class PlayerSpecificUniverse(private val originalUniverse: Universe, private val viewingPlayer: Player) {

    fun getPlanets(): List<PlayerSpecificPlanet> {
        return originalUniverse.getPlanets().map { PlayerSpecificPlanet(it, viewingPlayer) }
    }

    fun getHomePlanet(): PlayerSpecificPlanet? {
        val homePlanet = originalUniverse.getHomePlanetOf(viewingPlayer) ?: return null
        return PlayerSpecificPlanet(homePlanet, viewingPlayer)
    }

    fun getTravellingShipFormations(): Collection<ShipFormation> {
        return if (ManualTestingParameters.ENEMY_LOCATIONS_VISIBLE) originalUniverse.getTravellingShipFormations()
        else originalUniverse.getTravellingShipFormations()
                .filter { s -> s.getCommander() === viewingPlayer }
    }

    fun getPlanet(planetId: Int): PlayerSpecificPlanet {
        return PlayerSpecificPlanet(originalUniverse.getPlanetForId(planetId), viewingPlayer)
    }

    fun sendShips(shipCount: Int, originPlanetId: Int, destinationPlanetId: Int) {
        originalUniverse.sendShips(shipCount, originPlanetId, destinationPlanetId, viewingPlayer)
    }

    companion object {

        fun convert(universe: Universe, viewingPlayer: Player): PlayerSpecificUniverse {
            return PlayerSpecificUniverse(universe, viewingPlayer)
        }
    }
}
